pennylane/operation.py

```diff
@@ -474,45 +390,152 @@ def parameters(self):
+
+    @property
+    def grad_method(self):
+        """Gradient computation method.
+
+        * ``'A'``: analytic differentiation.
+        * ``'F'``: finite difference numerical differentiation.
+        * ``None``: the operation may not be differentiated.
+
+        Default is ``'F'``, or ``None`` if the Operation has zero parameters.
+        """
+        return None if self.num_params == 0 else 'F'
+
+    @property
+    def grad_recipe(self):
+        r"""Gradient recipe for the analytic differentiation method.
+
+        This is a list with one tuple per operation parameter. For parameter
+        :math:`k`, the tuple is of the form :math:`(c_k, s_k)`, resulting in
+        a gradient recipe of
+
+        .. math:: \frac{\partial}{\partial\phi_k}O = c_k\left[O(\phi_k+s_k)-O(\phi_k-s_k)\right].
+
+        If this property returns ``None``, the default gradient recipe
+        :math:`(c_k, s_k)=(1/2, \pi/2)` is assumed for every parameter.
+        """
+        return self._grad_recipe
+
+    @property
+    def generator(self):
+        r"""Generator of the operation.
+
+        A length-2 list ``[generator, scaling_factor]``, where
+
+        * ``generator`` is an existing PennyLane
+          operation class or :math:`2\times 2` Hermitian array
+          that acts as the generator of the current operation
+
+        * ``scaling_factor`` represents a scaling factor applied
+          to the generator operation
+
+        For example, if :math:`U(\theta)=e^{i0.7\theta \sigma_x}`, then
+        :math:`\sigma_x`, with scaling factor :math:`s`, is the generator
+        of operator :math:`U(\theta)`:
+
+        .. code-block:: python
+
+            generator = [PauliX, 0.7]
+
+        Default is ``[None, 1]``, indicating the operation has no generator.
+        """
+        return [None, 1]
+
+    @grad_recipe.setter
+    def grad_recipe(self, value):
+        """Setter for the grad_recipe property"""
+        self._grad_recipe = value
+
+    def __init__(self, *args, wires=None, do_queue=True):
+
+        # check the grad_method validity
+        if self.par_domain == 'N':
+            assert self.grad_method is None, 'An operation may only be differentiated with respect to real scalar parameters.'
+        elif self.par_domain == 'A':
+            assert self.grad_method in (None, 'F'), 'Operations that depend on arrays containing free variables may only be differentiated using the F method.'
+
+        # check the grad_recipe validity
+        if self.grad_method == 'A':
+            if self.grad_recipe is None:
+                # default recipe for every parameter
+                self.grad_recipe = [None] * self.num_params
+            else:
+                assert len(self.grad_recipe) == self.num_params, 'Gradient recipe must have one entry for each parameter!'
+        else:
+            assert self.grad_recipe is None, 'Gradient recipe is only used by the A method!'
+
+        super().__init__(*args, wires=wires, do_queue=do_queue)
```

Comment:
The grad recipe validation code is typically meant to be run just once, when the module is imported and the classes are created (as evidenced by the asserts), and validates the "static" properties of Operation subclasses, right?
If it can be put in the setter instead of `__init__` that would be ideal. Just check that all the class attributes that the code depends on (like `grad_method`) are set in the class definition before `grad_recipe`.

Reply:
Yep, sorry, I misread!